source/lac/scalapack.cc

```diff
@@ -0,0 +1,944 @@
+    active = true;
+
+  // Create an auxiliary communicator which has root and all inactive cores
+  // Assume that inactive cores start with id=n_process_rows*n_process_columns
+  Assert (active || this_mpi_process >= n_process_rows*n_process_columns,
+          ExcInternalError());
+
+  std::vector<int> inactive_with_root_ranks;
+  inactive_with_root_ranks.push_back(0);
+  for (int i = n_process_rows*n_process_columns; i < n_mpi_processes; ++i)
+    inactive_with_root_ranks.push_back(i);
+
+  // Get the group of processes in mpi_communicator
+  int ierr = 0;
+  MPI_Group all_group;
+  ierr = MPI_Comm_group(mpi_communicator, &all_group);
+  AssertThrowMPI(ierr);
+
+  // Construct the group containing all ranks we need:
+  MPI_Group inactive_with_root_group;
+  const int n = inactive_with_root_ranks.size();
+  ierr = MPI_Group_incl(all_group,
+                        n, &inactive_with_root_ranks[0],
+                        &inactive_with_root_group);
+  AssertThrowMPI(ierr);
+
+  // Create the communicator based on the group
+  // Note that, on most cores the communicator will be MPI_COMM_NULL
+  // FIXME: switch to MPI_Comm_create_group for MPI-3 so that only processes within the to-be subgroup call this
+  ierr = MPI_Comm_create(mpi_communicator, inactive_with_root_group,
+                         &mpi_communicator_inactive_with_root);
+  AssertThrowMPI(ierr);
+
+  MPI_Group_free(&all_group);
+  MPI_Group_free(&inactive_with_root_group);
+
+  // Double check that the process with rank 0 in subgroup is active:
+#ifdef DEBUG
+  if (mpi_communicator_inactive_with_root != MPI_COMM_NULL)
+    {
+      int subgroup_rank = -1, subgroup_size = -1;
+      MPI_Comm_rank(mpi_communicator_inactive_with_root, &subgroup_rank);
+      MPI_Comm_size(mpi_communicator_inactive_with_root, &subgroup_size);
+      if (subgroup_rank == 0)
+        Assert (active, ExcInternalError());
+    }
+#endif
+}
+
+
+
+ProcessGrid::ProcessGrid(MPI_Comm mpi_comm,
+                         const std::pair<unsigned int,unsigned int> &matrix_dimensions,
+                         const std::pair<unsigned int,unsigned int> &block_sizes)
+  :
+  ProcessGrid(mpi_comm,
+              choose_the_processor_grid(mpi_comm, matrix_dimensions.first, matrix_dimensions.second,
+                                        block_sizes.first, block_sizes.second) )
+{}
+
+
+
+ProcessGrid::~ProcessGrid()
+{
+  if (active)
+    Cblacs_gridexit(blacs_context);
+
+  MPI_Comm_free(&mpi_communicator_inactive_with_root);
+}
+
+
+
+unsigned int ProcessGrid::get_process_grid_rows() const
+{
+  return n_process_rows;
+}
+
+
+
+unsigned int ProcessGrid::get_process_grid_columns() const
+{
+  return n_process_columns;
+}
+
+
+
+template <typename NumberType>
+void ProcessGrid::send_to_inactive(NumberType *value, const int count) const
+{
+  Assert (count>0, ExcInternalError());
+  if (mpi_communicator_inactive_with_root != MPI_COMM_NULL)
+    {
+      const int ierr =
+        MPI_Bcast(value,count,MPI_DOUBLE,
```

Comment:
You need to choose the data type that matches `NumberType` -- it's not always going to be `MPI_DOUBLE`. There are functions that make this translation in `base/mpi.h`.